File: testSerial.py
#!/usr/bin/env python3
import serial
import time
import sys

def test_serial():
    try:
        # 打开串口
        ser = serial.Serial('/dev/ttyACM0', 115200, timeout=1)
        print(f"串口已打开: {ser.port}")
        print(f"波特率: {ser.baudrate}")
        print(f"超时: {ser.timeout}")
        
        # 等待串口稳定
        time.sleep(0.1)
        
        # 要发送的数据
        data = bytes([0x02, 0x01, 0x01])
        print(f"准备发送数据: {' '.join([f'0x{b:02X}' for b in data])}")
        
        # 发送数据
        bytes_written = ser.write(data)
        print(f"已发送 {bytes_written} 字节")
        
        # 强制刷新缓冲区
        ser.flush()
        print("缓冲区已刷新")
        
        # 等待一下，看是否有返回数据
        time.sleep(0.1)
        if ser.in_waiting > 0:
            response = ser.read(ser.in_waiting)
            print(f"收到响应: {' '.join([f'0x{b:02X}' for b in response])}")
        else:
            print("没有收到响应数据")
            
        # 关闭串口
        ser.close()
        print("串口已关闭")
        
    except serial.SerialException as e:
        print(f"串口错误: {e}")
        return False
    except Exception as e:
        print(f"其他错误: {e}")
        return False
    
    return True

if __name__ == "__main__":
    print("=== 串口测试脚本 ===")
    print("测试端口: /dev/ttyACM0")
    print("发送数据: 0x02 0x01 0x00")
    print("==================")
    
    if test_serial():
        print("测试完成")
    else:
        print("测试失败")
        sys.exit(1)

File: src/serial_pack/src/testSerial.cpp
#include <ros/ros.h>
#include <serial/serial.h>
#include <unistd.h>

int main(int argc, char **argv) {
  ros::init(argc, argv, "debug_test");

  try {
    ROS_INFO("=== Serial Debug Test ===");

    // Create serial object with parameters
    serial::Serial ser("/dev/ttyACM0", 115200,
                       serial::Timeout::simpleTimeout(1000));

    ROS_INFO("Port: %s", ser.getPort().c_str());
    ROS_INFO("Baudrate: %d", ser.getBaudrate());

    // Open serial port (already created with parameters)
    ROS_INFO("Opening serial port...");
    if (!ser.isOpen()) {
      ser.open();
    }

    ROS_INFO("Serial port is open: %s", ser.isOpen() ? "YES" : "NO");

    if (!ser.isOpen()) {
      ROS_ERROR("Failed to open serial port!");
      return 1;
    }

    // Wait for stabilization
    ROS_INFO("Waiting for port stabilization...");
    usleep(200000); // 200ms

    // Check initial status
    ROS_INFO("Available bytes: %zu", ser.available());

    // Prepare data
    std::vector<uint8_t> data = {0x02, 0x01, 0x01};
    ROS_INFO("Preparing to send %zu bytes: 0x02 0x01 0x00", data.size());

    // Send data
    ROS_INFO("Sending data...");
    size_t sent = ser.write(data);
    ROS_INFO("write() returned: %zu bytes", sent);

    if (sent == 0) {
      ROS_ERROR("Send failed! Returned 0 bytes");
    } else if (sent != data.size()) {
      ROS_WARN("Partial send: expected %zu, actual %zu", data.size(), sent);
    } else {
      ROS_INFO("Send successful!");
    }

    // Flush buffer
    ROS_INFO("Flushing buffer...");
    ser.flush();

    // Check status after flush
    ROS_INFO("Available bytes after flush: %zu", ser.available());
    ROS_INFO("Serial port still open: %s", ser.isOpen() ? "YES" : "NO");

    // Wait for response
    usleep(100000); // 100ms
    if (ser.available() > 0) {
      ROS_INFO("Received %zu bytes response", ser.available());
    } else {
      ROS_INFO("No response");
    }

    // Close serial port
    ser.close();
    ROS_INFO("Serial port closed");

  } catch (serial::IOException &e) {
    ROS_ERROR("Serial IO exception: %s", e.what());
    return 1;
  } catch (std::exception &e) {
    ROS_ERROR("Other exception: %s", e.what());
    return 1;
  }

  ROS_INFO("Test completed");
  return 0;
}